x-pack/plugins/lens/public/utils/native_renderer/native_renderer.tsx

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useEffect, useRef } from 'react';
+
+export interface NativeRendererProps<T> {
+  render: (domElement: Element, props: T) => void;
+  actualProps: T;
+  tag?: string;
+  children?: never;
+}
+
+function isShallowDifferent<T>(a: T, b: T): boolean {
+  for (const key in a) {
+    if (!(key in b) || a[key] !== b[key]) {
+      return true;
+    }
+  }
+  for (const key in b) {
+    if (!(key in a) || a[key] !== b[key]) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+export function NativeRenderer<T>({ render, actualProps, tag }: NativeRendererProps<T>) {
+  const elementRef = useRef<Element | null>(null);
+  const propsRef = useRef<T | null>(null);
+
+  function renderAndUpdate(element: Element) {
+    elementRef.current = element;
+    propsRef.current = actualProps;
+    render(element, actualProps);
+  }
+
+  useEffect(
+    () => {
+      if (elementRef.current && isShallowDifferent(propsRef.current, actualProps)) {
+        renderAndUpdate(elementRef.current);
+      }
+    },
+    [actualProps]
```

Comment:
I think we can get rid of `isShallowDifferent` altogether, and simplify the NativeRenderer to just be this:

```tsx
export function NativeRenderer<T>({ render, actualProps, tag }: NativeRendererProps<T>) {
  const elementRef = useRef<Element | null>(null);

  const rerender = () => {
    if (elementRef.current) {
      render(elementRef.current, actualProps);
    }
  };

  // Passing Object.values as the second argument here eliminates
  // the need for our isShallowDifferent function, as that is handled by React.
  // the order of the array returned by Object.values should be stable, based
  // on how we're creating it.
  useEffect(rerender, Object.values(actualProps));

  return React.createElement(tag || 'div', {
    ref: element => {
      if (element && element !== elementRef.current) {
        elementRef.current = element;
        rerender();
      }
    },
  });
}

```

Reply:
Heh, that's a nice idea. However I'd prefer having a more robust comparison function in place - it's entirely possible there will be ways to build props objects in the future which will yield different key orders. React has a compare function, but importing it directly is deprecated, so I just improved the existing implementation.